### ISSUE TITLE
Mid-generation pacing instability on a subset of long-form prose inputs

### ISSUE THREAD

**Issue by vdruts** [k2-fsa/OmniVoice#64]
First — thank you for OmniVoice. The model is genuinely excellent: voice cloning quality is remarkable for a 0.8B model, load times are fast, and it has been the primary TTS engine for a multi-voice production system we've been running for several weeks. Filing this as a real production bug report, not a drive-by.

## The symptom

On *some* longer text inputs, generation suffers from **mid-clip pacing acceleration**: the model starts at a normal pace, then partway through speeds up, and the speech starts to sound like it's racing — stumbling, tripping, clipping syllables. Listeners get pulled out of whatever is being said because the delivery becomes unnatural.

It is **inconsistent**. The same model with the same config produces mostly-good output on most inputs, then occasionally produces a clearly broken clip. The triggers are not obvious from the input alone.

## Observation from a 6-clip reproduction set

We built a reproduction harness using real production text from our system: six representative inputs across two voice profiles, with the exact production config. Listening blind:

- **3 of 6** sounded clean and natural — including longer 100+ word inputs and one with extensive numerical content (dollar amounts, percentages, multipliers like `$2.2M`, `15%`, `27x`).
- **2 of 6** were mostly fine but had a few words/phrases mid-clip where pacing tripped.
- **1 of 6** was clearly broken — racing, stumbling, unnatural enough that the message would not be usable in production.

The clearly-broken one was a long reflective prose passage (~150 words, ~45s of audio) characterized by:

- Flowing explanatory/philosophical tone, no strong sentence-rhythm anchors
- Few short emphatic sentences breaking up the long ones
- Several proper nouns and capitalized noun phrases
- **Not noticeably different in token count, vocabulary edge cases, or punctuation density** from the inputs that sounded fine

Importantly, the well-rendered inputs included content with what we initially expected would be tokenizer stress points (currency, percentages, day-name lists, em dashes, time references like \`3:30 PM\`, contractions). Those all produced clean output. So the failure mode does **not** appear to be driven by simple text-content edge cases.

## What we've ruled out

- **Sample rate mismatch.** Verified \`model.sampling_rate == 24000\`. Our pipeline writes WAV at 24000, ffmpeg upsamples for Opus output. Duration-preserving end-to-end.
- **Config drift between environments.** The same \`OmniVoiceGenerationConfig(num_step=40, guidance_scale=2.0, denoise=True, preprocess_prompt=True, postprocess_output=True)\` in both our test harness and our production server. Same \`create_voice_clone_prompt(ref_audio=..., ref_text=None)\`. Same model weights.
- **Speed parameter.** Passing \`speed=0.85\`–\`0.92\` slows the *overall* clip but does not eliminate the mid-generation acceleration on the broken clips.
- **Tokenizer stress on numbers/symbols.** Directly tested and ruled out as primary cause.

## What we suspect but can't yet confirm

- **Reference audio sensitivity.** Across two different reference clips of the same speaker — one ~12 seconds, one ~19 seconds — instability presents differently. We can't yet say cleanly whether this is reference-driven, content-driven, or an interaction, but it does not appear independent of the reference.
- **Long-sequence drift.** Clips that exhibit the bug are typically in the 40–50s output range. Shorter outputs (<25s) almost never exhibit it in our testing.
- **Diffusion sampler instability on certain prosodic patterns.** Speculating, but the failure being both content-dependent and reference-dependent points at something in how the LM/diffusion sampler maintains consistency across the sequence — possibly losing positional/rhythmic anchoring on flowing prose without strong syntactic boundaries.

## Reproduction config

\`\`\`python
import torch
from omnivoice import OmniVoice, OmniVoiceGenerationConfig

model = OmniVoice.from_pretrained(
    "k2-fsa/OmniVoice",
    device_map="cuda",
    dtype=torch.float16,
    load_asr=True,
)

prompt = model.create_voice_clone_prompt(
    ref_audio="path/to/your/reference.mp3",  # 15-20s clean speech
    ref_text=None,  # Whisper auto-transcribes
)

gen_config = OmniVoiceGenerationConfig(
    num_step=40,
    guidance_scale=2.0,
    denoise=True,
    preprocess_prompt=True,
    postprocess_output=True,
)

# Long flowing prose — the kind of input that sometimes triggers the bug
text = (
    "The biggest mistake I see in the field is treating tooling as a "
    "feature, not as a methodology. The team I worked with last year "
    "showed me that the platform isn't a product, it's a workflow "
    "system wrapped in a model. We rebuilt the offering around three "
    "principles: reduce effort, remove repetition, multiply the "
    "deliverable. That isn't we have automation. That's we compress "
    "the path from input to outcome. The pricing model isn't billable "
    "hours, it's a repeatable pattern that compounds. If you are still "
    "charging by the hour, you are trading time for money and you've "
    "already lost. The frame that changed everything was this: stop "
    "selling what you do and start selling the transformation."
)

audio = model.generate(
    text=text,
    voice_clone_prompt=prompt,
    generation_config=gen_config,
    speed=0.9,  # partial mitigation, does not fully eliminate
)
\`\`\`

System: Windows 11, consumer NVIDIA GPU, fp16. Same behavior observed at \`num_step=32, 40, 48\`.

## Questions for the maintainers

1. **Is there known content/length sensitivity** in the current 0.8B release that we should be working around?
2. **Are there generation parameters not exposed in \`OmniVoiceGenerationConfig\`** that affect long-sequence stability? Things like position temperature, sampler temperature, classifier-free-guidance variations, or any stability/quality tradeoff knob?
3. **Recommended reference audio properties** — optimal length, content type (declarative vs conversational), recording quality? We've observed that reference choice meaningfully affects output stability and would value any guidance on choosing references that maximize stability.
4. **Recommended text preprocessing** — should we be expanding numbers and abbreviations ourselves, normalizing certain punctuation, or chunking long inputs at sentence boundaries before passing to \`generate()\`?
5. **Larger model on the roadmap?** If a bigger checkpoint is planned, would it likely improve long-form stability?

We have invested significantly in building around OmniVoice and would much rather diagnose and work around this than swap engines. Happy to run additional tests with any configs you'd like us to try, or share private audio samples directly if that helps debugging.

Thank you again for the model.

**Comment by zhu-han** [comment]
Hi, I can't reproduce the issue you mentioned from the text you provided. I’ll need to see the actual problematic case to understand and fix it properly. Could you share a reproducible example (including the ref_audio if used) and the corresponding bad audio?
